contracts/Penalizer.sol

```diff
@@ -26,8 +26,30 @@ contract Penalizer is IPenalizer{
 
     }
 
-    modifier relayManagerOnly(IRelayHub hub) {
-        require(hub.isRelayManagerStaked(msg.sender), "Unknown relay manager");
+    mapping( bytes32=>bool ) public commits;
+
+    /**
+     * any sender can call "commit(keccak(encodedPenalizeFunction))", to make sure
+     * no-one can front-run it to claim this penalization
+     */
+    function commit(bytes32 commitHash) public {
+        bytes32 idx=keccak256(abi.encodePacked(commitHash,msg.sender));
+        commits[idx] = true;
+    }
+
+    function _commitRevealOrRelayManager(IRelayHub hub) internal {
+        bytes32 requestHash = keccak256(msg.data);
+        bytes32 idx=keccak256(abi.encodePacked(requestHash,msg.sender));
+
+        if ( commits[idx] ) {
```

Comment:
There is no requirement to have a delay in this method. This means that the frontrunner can totally put bot commit and reveal in the same block and still outrun the actual 'committer'.